docs/docs/reconciliation.md

```diff
@@ -0,0 +1,147 @@
+
+Any componts below the root will also get unmounted and have their state destroyed. For example, when diffing:
+
+```xml
+<div>
+  <Counter />
+</div>
+
+<span>
+  <Counter />
+</span>
+```
+
+This will destroy the old `Counter` and remount a new one.
+
+### DOM Elements Of The Same Type
+
+When comparing two React DOM elements of the same type, React looks at the attributes of both, keeps the same underlying DOM element, and only updates the changed attributes. For example:
+
+```xml
+<div className="before" title="stuff" />
+
+<div className="after" title="stuff" />
+```
+
+When converting between these two elements, React knows to only modify the `className`.
+
+When updating `style`, React also knows to update only the properties that changed. For example:
+
+```xml
+<div style={{'{{'}}color: 'red', fontWeight: 'bold'}} />
+
+<div style={{'{{'}}color: 'green', fontWeight: 'bold'}} />
+```
+
+When converting between these two elements, React knows to only modify the `color` style, not the `fontWeight`.
+
+After handling the DOM element, React then recurses on the children.
+
+### Component Elements Of The Same Type
+
+When a component updates, the instance stays the same, so that state is maintained across renders. React takes all the attributes from the new component element and calls `componentWillReceiveProps()` and `componentWillUpdate()` on the underlying instance.
+
+Next, the `render()` method is called and the diff algorithm recurses on the previous result and the new result.
+
+### Recursing On Children
+
+By default, when recursing on the children of a DOM element, React just iterates over both lists of children at the same time and generates a mutation whenever there's a difference.
+
+For example, when adding an element at the end of the children, converting between these two trees works well:
+
+```xml
+<ul>
+  <li>first</li>
+  <li>second</li>
+</ul>
+
+<ul>
+  <li>first</li>
+  <li>second</li>
+  <li>third</li>
+</ul>
+```
+
+React will match the two `<li>first</li>` trees, match the two `<li>second</li>` trees, and then insert the `<li>third</li>` tree.
+
+Inserting an element at the beginning has worse performance. For example, converting between these two trees works poorly:
+
+```xml
+<ul>
+  <li>Duke</li>
+  <li>Villanova</li>
+</ul>
+
+<ul>
+  <li>Connecticut</li>
+  <li>Duke</li>
+  <li>Villanova</li>
+</ul>
+```
+
+React will mutate every child instead of realizing it can keep the `<li>Duke</li>` and `<li>Villanova</li>` subtrees intact. This inefficiency can be a problem.
+
+### Keys
+
+In order to solve this issue, React supports a `key` attribute. When children have keys, React uses the key to match children in the original tree with children in the subsequent tree. For example, adding a `key` to our inefficient example above can make the tree conversion efficient:
+
+```xml
+<ul>
+  <li key={'2015'}>Duke</li>
```

Comment:
The notation is weird. Let's just use string attributes here (="") and then add a single line example with the words like "Most often, the key would come from your data, such as an ID:" (and the example does something like lib key={todo.id})